proof/invariant-abstract/RISCV64/ArchKernelInit_AI.thy

```diff
@@ -237,81 +275,197 @@ lemma kernel_mapping_slots_0x100[simp]:
   apply (simp add: pptrBaseOffset_def paddrBase_def)
-  apply (simp add: pt_bits_left_def bit_simps level_defs)
+  apply (simp add: pt_bits_left_def bit_simps level_defs elf_index_value toplevel_bits_def)
   apply (rule conjI)
    apply (rule is_aligned_add)
     apply (simp add: mask_def)
    apply (simp add: pptrBase_def canonical_bit_def is_aligned_def)
-  apply (simp add: pptr_base_def)
+  apply (simp add: pptr_base_def kernel_window_bits_def)
   apply (simp add: pptrBase_def neg_mask_le_high_bits flip: NOT_mask)
   apply (subst word_plus_and_or_coroll; simp add: canonical_bit_def word_size mask_def)
   apply word_bitwise
   apply clarsimp
   done
 
-lemma elf_window_1M:
-  "\<lbrakk> kernel_elf_base \<le> vref; vref < kernel_elf_base + (1 << 20) \<rbrakk> \<Longrightarrow>
-    table_index (pt_slot_offset max_pt_level riscv_global_pt_ptr vref) = 0x1FE"
-  apply (simp add: table_index_riscv_global_pt_ptr)
-  apply (simp add: bit_simps kernel_elf_base_def kernelELFBase_def kernelELFPAddrBase_def
-                   Kernel_Config.physBase_def)
+lemma is_page_aligned_physBase: (* FIXME MK: move to Kernel_Config_Lemmas *)
+  "is_aligned physBase pageBits"
+  by (simp add: Kernel_Config.physBase_def pageBits_def is_aligned_def)
+
+lemma is_page_aligned_kernelELFPAddrBase: (* FIXME MK: does not mention physBase, but could be moved up *)
+  "is_aligned kernelELFPAddrBase pageBits"
+  unfolding kernelELFPAddrBase_def
+  by (fastforce intro!: is_aligned_add is_page_aligned_physBase
+                simp: kernelELFPAddrBase_def pageBits_def is_aligned_def)
+
+(* FIXME MK: move *)
+lemma aligned_mask_le_mask_minus:
+  "\<lbrakk> is_aligned x m; m \<le> n; n < LENGTH('a)\<rbrakk> \<Longrightarrow> x && mask n \<le> mask n - mask m" for x :: "'a::len word"
+  by (metis and_mask_less' is_aligned_after_mask is_aligned_neg_mask_eq'
+           mask_2pm1 mask_sub neg_mask_mono_le word_less_sub_le)
+
+(* FIXME MK: rename, move *)
+lemma aligned_mask_plus_bounded:
+  "\<lbrakk> is_aligned x m; m < n; n < LENGTH('a) \<rbrakk> \<Longrightarrow> (x && mask n) + 2^m \<le> 2^n" for x :: "'a::len word"
+  by (metis add_mask_fold and_mask_less' is_aligned_add_step_le is_aligned_after_mask is_aligned_mask
+            leD leI order_less_imp_le t2n_mask_eq_if word_less_sub_1)
+
+lemma kernelELF_plus_page:
+  "((kernelELFPAddrBase && mask toplevel_bits) + 2^pageBits) \<le> 2^toplevel_bits"
+  by (fastforce intro: aligned_mask_plus_bounded is_page_aligned_kernelELFPAddrBase
+                simp: toplevel_bits_value bit_simps)
+
+lemma elf_window_4k:
+  "\<lbrakk> kernel_elf_base \<le> vref; vref < kernel_elf_base + (1 << pageBits) \<rbrakk> \<Longrightarrow>
+    table_index (pt_slot_offset max_pt_level riscv_global_pt_ptr vref) = elf_index"
+  using is_page_aligned_kernelELFPAddrBase
+  apply (simp add: table_index_riscv_global_pt_ptr elf_index_value toplevel_bits_value)
+  apply (unfold kernel_elf_base_def kernelELFBase_def bit_simps)
+  apply (simp add: add_ac)
```

Comment:
unfold + simp can be combined into one simp?

Reply:
The best I managed without spending a lot more time on this is
```
  using is_page_aligned_kernelELFPAddrBase
  apply (simp add: table_index_riscv_global_pt_ptr elf_index_value toplevel_bits_value
                   kernel_elf_base_def kernelELFBase_def)
  apply (simp add: bit_simps add_ac)
```
The simplifier makes things more complicated if `bit_simps` is used too early.

Reply:
That sounds familiar, I think I explicitly tried to single-step the unfold because bit_simps and add_ac were interacting somehow